# 💻 **Introduction**

Sebelum sebuah data dapat dianalisis atau digunakan untuk mengambil keputusan kebijakan tentu data tersebut perlu untuk diekstraksi terlebih dahulu. <b>Ekstraksi data</b> adalah sebuah proses pengumpulan atau pengambilan jenis data yang berbeda dari berbagai sumber, banyak di antaranya mungkin tidak terorganisir dengan baik atau tidak terstruktur sama sekali. Proses ini memungkinkan untuk mengkonsolidasikan, memproses, dan menyempurnakan data sehingga nantinya dapat disimpan di lokasi terpusat untuk diubah atau diproses lebih lanjut.

Python sebagai salah satu <i>tools</i> yang <i>powerful</i> untuk pengolahan data menyediakan berbagai <i>library</i> untuk mendukung proses ekstraksi data. Salah satu library paling banyak digunakan adalah <b>pandas</b>.

Pandas memungkinkan untuk membaca data dari berbagai jenis ekstensi seperti data yang bersumber dari database (sql), <i>excel</i> (<b>.xls</b> atau <b>.xlsx</b>), <i>separated value file</i> (<b>.csv</b>, <b>.tsv</b>, dsb) lalu merubahnya menjadi satu format struktur data yang sama yang disebut <b>DataFrame</b>. DataFrame inilah yang nantinya dapat kita lakukan proses <b>Data Manipulation</b><br><br>

<p align="center">
<img src="https://drive.google.com/uc?export=view&id=18NW3gMMTJjkwwpZX5TXi9DTteXxGAoxa" width="600" height="300">
<p align="center"><b>Gambar 1. Proses Ekstraksi Data dengan <i>pandas</i></b></p>

<br><br>
<p align="right"><sup>docs : <i><a href="https://shorturl.at/xyZ69">https://pandas.pydata.org/docs/user_guide/io.html</i></sup></a></p>


## 🏃🏻‍♂️ **1. First Step : Pandas**

Untuk dapat menggunakan library pandas pada python, perlu untuk mengimpor library tersebut dengan cara mengetikan sintaks berikut

<p align="center"><code><b>import pandas as pd</b></code></p>

lalu jalankan (<i>run</i>) program.

<br><br>
<p align="left"><sup>docs : <i><a href="https://shorturl.at/kzG06">https://pandas.pydata.org/docs/getting_started/install.html</i></sup></a></p>

### 👨‍💻 Task 1

Ketikan kode `import pandas as pd` pada prompt dibawah ini lalu jalankan

In [ ]:
# Write your code below here & exec!
import pandas as pd

# 🥨 **Struktur Data Pandas**

* Terdapat dua struktur data utama pada library pandas yang disebut dengan Series dan DataFrame.
<p align="center">
<img src="https://datagy.io/wp-content/uploads/2022/12/Understanding-Pandas-Series-and-DataFrames-1024x576.png" width="550" height="289">
<p align="center"><b>Gambar 2. Series VS DataFrame</b></p>
<br>

* Gabungan beberapa Series yang mempunyai banyak baris yang sama dapat membentuk sebuah DataFrame atau berlaku sebaliknya pada sebuah DataFrame dapat dipecah (di-<i>slicing</i> satu kolom) membentuk sebuah Series.<br><br>

<p align="right"><sup>docs : <i><a href="https://shorturl.at/dezQY">http://pandas.pydata.org/docs/getting_started/intro_tutorials/01_table_oriented.html</i></sup></a></p>

##🛺 **2. Series**

* <a href="">Series</a> merupakan kolom dari sebuah tabel yang bersifat homogen atau mempunyai elemen dengan tipe data yang sejenis.

* Untuk membuat Series gunakan sintaks berikut

<p align="center"><code><b>pd.Series(data)</b></code></p>

<ul>
  data dapat berupa <i>array-like</i>, <i>Iterable</i>, <i>dict</i>, atau <i>scalar</i>
</ul>

<p align="right"><sup>docs : <i><a href="https://shorturl.at/gmtG2">https://pandas.pydata.org/docs/reference/api/pandas.Series.html</i></sup></a></p>



### 👨‍💻 Task 2 - A

<i>Age of Accounts Receivable</i> (Aging AR) adalah istilah yang merujuk pada lamanya piutang dagang yang masih belum tertagih pada suatu periode tertentu. DQTech mecatat distribusi Aging AR tersaji dari list data sebagai berikut \[78, 91, 40, 50, 95\].

Dari data tersebut buatlah sebuah Series yang disimpan pada variabel dengan nama `ar_aging` lalu tampilkan outputnya!

In [ ]:
# Inisialisasi data
data_age_of_AR = [78, 91, 40, 50, 95]

# Write your code below here & exec!
# konversi ke series dr list
ar_aging = pd.Series(data_age_of_AR)

# Menampilkan hasil
display(ar_aging)

0    78
1    91
2    40
3    50
4    95
dtype: int64

In [ ]:
print(ar_aging)

0    78
1    91
2    40
3    50
4    95
dtype: int64


### 👨‍💻 Task 2 - B

Dari data Aging AR yang tercatat, akan diberi detail data sebagai berikut
* Cust 1 memiliki Aging AR sebesar 78
* Cust 2 memiliki Aging AR sebesar 91
* Cust 3 memiliki Aging AR sebesar 40
* Cust 4 memiliki Aging AR sebesar 50
* Cust 5 memiliki Aging AR sebesar 95

Dari data tersebut buatlah sebuah Series yang disimpan pada variabel dengan nama `detail_ar_aging` lalu tampilkan outputnya!

In [ ]:
# Inisialisasi data
data_age_of_AR = {
    'Cust1' : 78,
    'Cust2' : 91,
    'Cust3' : 40,
    'Cust4' : 50,
    'Cust5' : 95
}

# Write your code below here & exec!
detail_ar_aging = pd.Series(data_age_of_AR)

# Tampilkan hasilnya
display(detail_ar_aging)

Cust1    78
Cust2    91
Cust3    40
Cust4    50
Cust5    95
dtype: int64

## 🚴🏻‍♂️ **3. DataFrame**

* <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html#pandas.DataFrame">DataFrame</a> adalah struktur data 2 dimensi yang terdiri dari baris dan kolom (seperti tabel) dan dapat menyimpan data dari berbagai jenis (termasuk karakter, bilangan bulat, bilangan real, data kategorikal, dan lainnya).

<p align="center">
<img src="https://www.w3resource.com/w3r_images/pandas-data-frame.svg" width="600" height="300">
<p align="center"><b>Gambar 2. DataFrame Pandas</b></p>
<br>

* <a href="https://pandas.pydata.org/docs/reference/api/pandas.Series.html#pandas.Series">Series</a> merupakan kolom dari sebuah tabel yang bersifat homogen atau mempunyai elemen dengan tipe data yang sejenis.


<p align="center">
<img src="https://datagy.io/wp-content/uploads/2022/12/Understanding-Pandas-Series-and-DataFrames-1024x576.png" width="550" height="300">
<p align="center"><b>Gambar 3. Series VS DataFrame</b></p>
<br>

* Gabungan beberapa Series yang mempunyai banyak baris yang sama dapat membentuk sebuah DataFrame atau berlaku sebaliknya pada sebuah DataFrame dapat dipecah (di-<i>slicing</i> satu kolom) membentuk sebuah Series.<br><br>




### 👨‍💻 Task 3

Data karyawan DQTech disajikan pada dictionary berikut

```
details = {
    'Name' : ['Aksara', 'Antara', 'Sendja', 'Sunyi'],
    'Age' : [23, 21, 22, 21],
    'Role' : ['Jr. Data Scientist', 'Data Engineer', 'Data Analyst', 'Data Scientist']
}
```

konversi dictionary tersebut menjadi DataFrame pandas

In [ ]:
# Import library
import pandas as pd

# Initialize Data
details = {
    'Name' : ['Aksara', 'Antara', 'Sendja', 'Sunyi'],
    'Age' : [23, 21, 22, 21],
    'Role' : ['Jr. Data Scientist', 'Data Engineer', 'Data Analyst', 'Data Scientist']
}

# Write your code below here & exec!
dqtech_employee = pd.DataFrame(details)

# Tampilkan data
display(dqtech_employee)

Name  Age                Role
0  Aksara   23  Jr. Data Scientist
1  Antara   21       Data Engineer
2  Sendja   22        Data Analyst
3   Sunyi   21      Data Scientist

In [ ]:
type(dqtech_employee)

pandas.core.frame.DataFrame

## ♖ **4. Parts Of DataFrame**

Untuk lebih memahami objek DataFrame, perlu diketahui bahwa DataFrame terdiri dari tiga komponen sebagai atribut berikut :

<ul>1️⃣ .values : Nilai - nilai yang tertera pada DataFrame</ul>
<ul>2️⃣ .columns : Nama kolom yang ada pada DataFrame</ul>
<ul>3️⃣ .index : Nama atau nomor index pada DataFrame</ul>



### 👨‍💻 Task 4

Pada DataFrame yang telah dibuat pada Task 3, coba buatkan list nama kolomnya!

In [ ]:
# Write your code below here & exec!
# DAPETIN NAMA KOLOMNYA

dqtech_employee.columns

Index(['Name', 'Age', 'Role'], dtype='object')

In [ ]:
# Write your code below here & exec!
# DAPETIN VALUENYA

dqtech_employee.values

array([['Aksara', 23, 'Jr. Data Scientist'],
       ['Antara', 21, 'Data Engineer'],
       ['Sendja', 22, 'Data Analyst'],
       ['Sunyi', 21, 'Data Scientist']], dtype=object)

In [ ]:
# Write your code below here & exec!
# DAPETIN VALUENYA

dqtech_employee.index

RangeIndex(start=0, stop=4, step=1)

---


# 📊 **Data Manipulation**

Data Manipulation adalah sebuah metode pengorganisasian data agar lebih mudah dibaca, terstruktur dan agar lebih mudah diolah untuk kebutuhan lain kedepannya. Beberapa hal yang biasa dilakukan saat data manipulation adalah sebagai berikut :


<ul>1️⃣ Column selections</ul>
<ul>2️⃣ Filter the data</ul>
<ul>3️⃣ Create a new column or change it</ul>
<ul>4️⃣ Drop a column</ul>
<ul>5️⃣ Apply custom functions</ul>
<ul>6️⃣ Join / Merge the data</ul>
<ul>7️⃣ etc</ul>


## ⚓️ **5. Data Extraction**

* Dalam konteks data extraction, Pandas dapat digunakan untuk membaca data dari berbagai sumber, seperti file CSV, Excel, SQL database, dan lainnya, dan mengubahnya menjadi DataFrames untuk analisis lebih lanjut.

* Beberapa fungsi dan metode Pandas yang umum digunakan untuk data extraction adalah sebagai berikut:

<ul>
<code><a href="https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html">pd.read_csv()</a></code>: Digunakan untuk membaca data dari file CSV dan mengonversinya menjadi DataFrame.

<code><a href="https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html">pd.read_excel()</a></code>: Membaca data dari file Excel dan membuat DataFrame dari lembar kerja atau tabel tertentu.

<code><a href="https://pandas.pydata.org/docs/reference/api/pandas.read_sql.html">pd.read_sql()</a></code>: Mengambil data dari basis data SQL menggunakan SQL query dan menyimpannya dalam bentuk DataFrame.
</ul>


### 👨‍💻 Task 5

Untuk melakukan proses ekstraksi data, lakukan instruksi berikut :     

1.   Persiapkan dataset SuperStore.xlsx
2.   Upload ke google colab dengan cara pilih File pada bagian kiri laman (icon 📁), kemudian klik kanan dan pilih upload
3.   Lakukan ekstraksi data excel (.xlsx) dengan library pandas, assign hasilnya ke sebuah variabel bernama <code>data_superstore</code>
4.   Tampilkan hasilnya






In [ ]:
# Import library
import pandas as pd

# Proses ekstraksi data
data_superstore = pd.read_excel('/content/Superstore.xlsx')

# Tampilkeun hasilnya
display(data_superstore)

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
0          1  CA-2016-152156 2016-11-08 2016-11-11    Second Class   
1          2  CA-2016-152156 2016-11-08 2016-11-11    Second Class   
2          3  CA-2016-138688 2016-06-12 2016-06-16    Second Class   
3          4  US-2015-108966 2015-10-11 2015-10-18  Standard Class   
4          5  US-2015-108966 2015-10-11 2015-10-18  Standard Class   
...      ...             ...        ...        ...             ...   
9989    9990  CA-2014-110422 2014-01-21 2014-01-23    Second Class   
9990    9991  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9991    9992  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9992    9993  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9993    9994  CA-2017-119914 2017-05-04 2017-05-09    Second Class   

     Customer ID     Customer Name    Segment        Country             City  \
0       CG-12520       Claire Gute   Consumer  United States        Henderson   
1       CG-12520       Claire Gute   Consumer  United States        Henderson   
2       DV-13045   Darrin Van Huff  Corporate  United States      Los Angeles   
3       SO-20335    Sean O'Donnell   Consumer  United States  Fort Lauderdale   
4       SO-20335    Sean O'Donnell   Consumer  United States  Fort Lauderdale   
...          ...               ...        ...            ...              ...   
9989    TB-21400  Tom Boeckenhauer   Consumer  United States            Miami   
9990    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9991    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9992    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9993    CC-12220      Chris Cortes   Consumer  United States      Westminster   

      ... Postal Code  Region       Product ID         Category Sub-Category  \
0     ...       42420   South  FUR-BO-10001798        Furniture    Bookcases   
1     ...       42420   South  FUR-CH-10000454        Furniture       Chairs   
2     ...       90036    West  OFF-LA-10000240  Office Supplies       Labels   
3     ...       33311   South  FUR-TA-10000577        Furniture       Tables   
4     ...       33311   South  OFF-ST-10000760  Office Supplies      Storage   
...   ...         ...     ...              ...              ...          ...   
9989  ...       33180   South  FUR-FU-10001889        Furniture  Furnishings   
9990  ...       92627    West  FUR-FU-10000747        Furniture  Furnishings   
9991  ...       92627    West  TEC-PH-10003645       Technology       Phones   
9992  ...       92627    West  OFF-PA-10004041  Office Supplies        Paper   
9993  ...       92683    West  OFF-AP-10002684  Office Supplies   Appliances   

                                           Product Name     Sales  Quantity  \
0                     Bush Somerset Collection Bookcase  261.9600         2   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2     Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3         Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                        Eldon Fold 'N Roll Cart System   22.3680         2   
...                                                 ...       ...       ...   
9989                             Ultra Door Pull Handle   25.2480         3   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...   91.9600         2   
9991                              Aastra 57i VoIP phone  258.5760         2   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"   29.6000         4   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...  243.1600         2   

      Discount    Profit  
0         0.00   41.9136  
1         0.00  219.5820  
2         0.00    6.8714  
3         0.45 -383.0310  
4         0.20    2.5164  
...        ...       ...  
9989      0.20    4.1028  
9990      0.00   15.6332  
9991      0.20   19.3932  
9992      0.

In [ ]:
# TAMPILIN KOLOM SECARA LENGKAP

# Import library
import pandas as pd
pd.set_option('display.max_columns', None)

# Proses ekstraksi data
data_superstore = pd.read_excel('/content/Superstore.xlsx')

# Tampilkeun hasilnya
display(data_superstore)

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
0          1  CA-2016-152156 2016-11-08 2016-11-11    Second Class   
1          2  CA-2016-152156 2016-11-08 2016-11-11    Second Class   
2          3  CA-2016-138688 2016-06-12 2016-06-16    Second Class   
3          4  US-2015-108966 2015-10-11 2015-10-18  Standard Class   
4          5  US-2015-108966 2015-10-11 2015-10-18  Standard Class   
...      ...             ...        ...        ...             ...   
9989    9990  CA-2014-110422 2014-01-21 2014-01-23    Second Class   
9990    9991  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9991    9992  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9992    9993  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9993    9994  CA-2017-119914 2017-05-04 2017-05-09    Second Class   

     Customer ID     Customer Name    Segment        Country             City  \
0       CG-12520       Claire Gute   Consumer  United States        Henderson   
1       CG-12520       Claire Gute   Consumer  United States        Henderson   
2       DV-13045   Darrin Van Huff  Corporate  United States      Los Angeles   
3       SO-20335    Sean O'Donnell   Consumer  United States  Fort Lauderdale   
4       SO-20335    Sean O'Donnell   Consumer  United States  Fort Lauderdale   
...          ...               ...        ...            ...              ...   
9989    TB-21400  Tom Boeckenhauer   Consumer  United States            Miami   
9990    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9991    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9992    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9993    CC-12220      Chris Cortes   Consumer  United States      Westminster   

           State  Postal Code Region       Product ID         Category  \
0       Kentucky        42420  South  FUR-BO-10001798        Furniture   
1       Kentucky        42420  South  FUR-CH-10000454        Furniture   
2     California        90036   West  OFF-LA-10000240  Office Supplies   
3        Florida        33311  South  FUR-TA-10000577        Furniture   
4        Florida        33311  South  OFF-ST-10000760  Office Supplies   
...          ...          ...    ...              ...              ...   
9989     Florida        33180  South  FUR-FU-10001889        Furniture   
9990  California        92627   West  FUR-FU-10000747        Furniture   
9991  California        92627   West  TEC-PH-10003645       Technology   
9992  California        92627   West  OFF-PA-10004041  Office Supplies   
9993  California        92683   West  OFF-AP-10002684  Office Supplies   

     Sub-Category                                       Product Name  \
0       Bookcases                  Bush Somerset Collection Bookcase   
1          Chairs  Hon Deluxe Fabric Upholstered Stacking Chairs,...   
2          Labels  Self-Adhesive Address Labels for Typewriters b...   
3          Tables      Bretford CR4500 Series Slim Rectangular Table   
4         Storage                     Eldon Fold 'N Roll Cart System   
...           ...                                                ...   
9989  Furnishings                             Ultra Door Pull Handle   
9990  Furnishings  Tenex B1-RE Series Chair Mats for Low Pile Car...   
9991       Phones                              Aastra 57i VoIP phone   
9992        Paper  It's Hot Message Books with Stickers, 2 3/4" x 5"   
9993   Appliances  Acco 7-Outlet Masterpiece Power Center, Wihtou...   

         Sales  Quantity  Discount    Profit  
0     261.9600         2      0.00   41.9136  
1     731.9400         3      0.00  219.5820  
2      14.6200         2      0.00    6.8714  
3     957.5775         5      0.45 -383.0310  
4      22.3680         2      0.20    2.5164  
...        ...       ...       ...       ...  
9989   25.2480         3      0.20    4.1028  
9990   91.9600         2      0.00   15.6332  
9991  258.5760 

## ⛑ **6. Inspecting Data Information**

* Untuk dapat menginspeksi informasi umum pada data dapat menggunakan sintaks berikut

<p align="center"><code><b>DataFrame.info()</b></code></p>

<ul>Ketika dijalankan, sintaks tersebut akan menghasilkan output informasi umum pada data seperti banyak kolom, nama kolom beserta tipe data dan jumlah data tak kosong (<i>non-null value</i>)</ul><br>

<p align="right"><sup>docs : <i><a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html</i></sup></a></p>


### 👨‍💻 Task 6

Inspeksi informasi umum pada data yang telah diekstraksi dan telah disimpan pada variabel <code>data_superstore</code>. Setelah itu coba uraikan informasi apa yang bisa anda dapatkan setelah melakukan inspeksi informasi umum pada <code>data_superstore</code>?

In [ ]:
# Write your code below here & exec!
data_superstore.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Row ID         9994 non-null   int64         
 1   Order ID       9994 non-null   object        
 2   Order Date     9994 non-null   datetime64[ns]
 3   Ship Date      9994 non-null   datetime64[ns]
 4   Ship Mode      9994 non-null   object        
 5   Customer ID    9994 non-null   object        
 6   Customer Name  9994 non-null   object        
 7   Segment        9994 non-null   object        
 8   Country        9994 non-null   object        
 9   City           9994 non-null   object        
 10  State          9994 non-null   object        
 11  Postal Code    9994 non-null   int64         
 12  Region         9994 non-null   object        
 13  Product ID     9994 non-null   object        
 14  Category       9994 non-null   object        
 15  Sub-Category   9994 n

## 🔱 **7. Data Selection**

* Untuk memilih kolom tertentu pada sebuah DataFrame dapat menggunakan sintaks berikut

<p align="center"><code><b>DataFrame.loc[baris : nama_kolom]</b></code></p>

* <code>baris</code>
<ul>
  <li>Gunakan simbol ( : ) untuk menampilkan semua data</li>
  <li>Gunakan simbol (m : n) untuk menampilkan data dari index ke - m sampai dengan index ke - n</li>
</ul>

* <code>nama_kolom</code>
<ul>
  <li>Gunakan string untuk mengakses satu kolom saja (<i>return : Series</i>)</li>
  <li>Gunakan list untuk mengakses lebih dari satu kolom  (<i>return : DataFrame</i>)</li>
</ul>


### 👨‍💻 Task 7 - A

Pada DataFrame <code>data_superstore</code>, simpan 10 data teratas ke dalam variabel **order_id** untuk kolom **Order_ID** saja. Kemudian tampilkan data tersebut!

In [ ]:
data_superstore.head(1)

Row ID        Order ID Order Date  Ship Date     Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   

  Customer Name   Segment        Country       City  ... Postal Code  Region  \
0   Claire Gute  Consumer  United States  Henderson  ...       42420   South   

        Product ID   Category Sub-Category                       Product Name  \
0  FUR-BO-10001798  Furniture    Bookcases  Bush Somerset Collection Bookcase   

    Sales  Quantity  Discount   Profit  
0  261.96         2       0.0  41.9136  

[1 rows x 21 columns]

In [ ]:
# Write your code below here & exec!
order_id = data_superstore.loc[0:9, 'Order ID']

display(order_id)

0    CA-2016-152156
1    CA-2016-152156
2    CA-2016-138688
3    US-2015-108966
4    US-2015-108966
5    CA-2014-115812
6    CA-2014-115812
7    CA-2014-115812
8    CA-2014-115812
9    CA-2014-115812
Name: Order ID, dtype: object

### 👨‍💻 Task 7 - B

Pada DataFrame <code>data_superstore</code>, tampilkan data dengan index ke-2 sampai dengan index ke-5 kolom **Order_ID**, **Sales**, **Profit** saja

In [ ]:
data_superstore['Order ID'][0:9]

0    CA-2016-152156
1    CA-2016-152156
2    CA-2016-138688
3    US-2015-108966
4    US-2015-108966
5    CA-2014-115812
6    CA-2014-115812
7    CA-2014-115812
8    CA-2014-115812
Name: Order ID, dtype: object

In [ ]:
# Write your code below here & exec!
col_selection = ['Order ID', 'Sales', 'Profit']

data_sales = data_superstore.loc[2:5][col_selection]

display(data_sales)

Order ID     Sales    Profit
2  CA-2016-138688   14.6200    6.8714
3  US-2015-108966  957.5775 -383.0310
4  US-2015-108966   22.3680    2.5164
5  CA-2014-115812   48.8600   14.1694

In [ ]:
# Write your code below here & exec!
col_selection = ['Order ID', 'Sales', 'Profit']

data_sales = data_superstore.loc[2:5, col_selection]

display(data_sales)

Order ID     Sales    Profit
2  CA-2016-138688   14.6200    6.8714
3  US-2015-108966  957.5775 -383.0310
4  US-2015-108966   22.3680    2.5164
5  CA-2014-115812   48.8600   14.1694

atau secara sederhana anda dapat menggunakan `DataFrame[nama_kolom]` untuk menghasilkan sebuah Series satu kolom dan `DataFrame[[nama_kolom1, nama_kolom2, ...]]` untuk menseleksi banyak kolom kedalam bentuk DataFrame

### 👨‍💻 Task 7 - C

Pada DataFrame <code>data_superstore</code>, simpan ke dalam variabel **order_id** untuk kolom **Order_ID** saja (gunakan seleksi sederhana). Kemudian tampilkan data tersebut!

In [ ]:
# Write your code below here & exec!
order_id = data_superstore['Order ID']

display(order_id)

0       CA-2016-152156
1       CA-2016-152156
2       CA-2016-138688
3       US-2015-108966
4       US-2015-108966
             ...      
9989    CA-2014-110422
9990    CA-2017-121258
9991    CA-2017-121258
9992    CA-2017-121258
9993    CA-2017-119914
Name: Order ID, Length: 9994, dtype: object

## 🎵 **8. Conditions**

* Salah satu teknik paling sederhana untuk menemukan bagian mana dari kumpulan data yang menarik adalah menemukan subset baris yang cocok dengan beberapa kriteria untuk difilter atau dipilih. Untuk melakukan hal ini cukup mudah yakni dengan menuliskan kondisi seperti berikut

<p align="center"><code><b>DataFrame.loc[conditions]</b></code></p>

* Conditions dapat dibentuk dari 1 atau lebih kriteria boolean dan dihubungkan dengan operator logika and (&), or (|) atau not (~)

<br>

docs : <i>https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html</i>

### 👨‍💻 Task 8 - A

Pada DataFrame `data_superstore` tampilkan semua record data yang terjadi di 'Los Angeles' saja

In [ ]:
# Write your code below here & exec!

# Assign conditions
conditions = (data_superstore['City'] == 'Los Angeles')

# Filter
data_LA = data_superstore.loc[conditions]

# Tampilkeun
display(data_LA)

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
2          3  CA-2016-138688 2016-06-12 2016-06-16    Second Class   
5          6  CA-2014-115812 2014-06-09 2014-06-14  Standard Class   
6          7  CA-2014-115812 2014-06-09 2014-06-14  Standard Class   
7          8  CA-2014-115812 2014-06-09 2014-06-14  Standard Class   
8          9  CA-2014-115812 2014-06-09 2014-06-14  Standard Class   
...      ...             ...        ...        ...             ...   
9976    9977  US-2016-103674 2016-12-06 2016-12-10  Standard Class   
9977    9978  US-2016-103674 2016-12-06 2016-12-10  Standard Class   
9978    9979  US-2016-103674 2016-12-06 2016-12-10  Standard Class   
9979    9980  US-2016-103674 2016-12-06 2016-12-10  Standard Class   
9986    9987  CA-2016-125794 2016-09-29 2016-10-03  Standard Class   

     Customer ID    Customer Name      Segment        Country         City  \
2       DV-13045  Darrin Van Huff    Corporate  United States  Los Angeles   
5       BH-11710  Brosina Hoffman     Consumer  United States  Los Angeles   
6       BH-11710  Brosina Hoffman     Consumer  United States  Los Angeles   
7       BH-11710  Brosina Hoffman     Consumer  United States  Los Angeles   
8       BH-11710  Brosina Hoffman     Consumer  United States  Los Angeles   
...          ...              ...          ...            ...          ...   
9976    AP-10720       Anne Pryor  Home Office  United States  Los Angeles   
9977    AP-10720       Anne Pryor  Home Office  United States  Los Angeles   
9978    AP-10720       Anne Pryor  Home Office  United States  Los Angeles   
9979    AP-10720       Anne Pryor  Home Office  United States  Los Angeles   
9986    ML-17410     Maris LaWare     Consumer  United States  Los Angeles   

      ... Postal Code  Region       Product ID         Category Sub-Category  \
2     ...       90036    West  OFF-LA-10000240  Office Supplies       Labels   
5     ...       90032    West  FUR-FU-10001487        Furniture  Furnishings   
6     ...       90032    West  OFF-AR-10002833  Office Supplies          Art   
7     ...       90032    West  TEC-PH-10002275       Technology       Phones   
8     ...       90032    West  OFF-BI-10003910  Office Supplies      Binders   
...   ...         ...     ...              ...              ...          ...   
9976  ...       90032    West  TEC-PH-10002496       Technology       Phones   
9977  ...       90032    West  OFF-FA-10003467  Office Supplies    Fasteners   
9978  ...       90032    West  OFF-BI-10003727  Office Supplies      Binders   
9979  ...       90032    West  OFF-BI-10002026  Office Supplies      Binders   
9986  ...       90008    West  TEC-AC-10003399       Technology  Accessories   

                                           Product Name    Sales  Quantity  \
2     Self-Adhesive Address Labels for Typewriters b...   14.620         2   
5     Eldon Expressions Wood and Plastic Desk Access...   48.860         7   
6                                            Newell 322    7.280         4   
7                        Mitel 5320 IP Phone VoIP phone  907.152         6   
8     DXL Angle-View Binders with Locking Rings by S...   18.504         3   
...                                                 ...      ...       ...   
9976                                       Cisco SPA301  249.584         2   
9977           Alliance Big Bands Rubber Bands, 12/Pack   13.860         7   
9978  Avery Durable Slant Ring Binders With Label Ho...   13.376         4   
9979                  Ibico Recycled Linen-Style Covers  437.472        14   
9986  Memorex Mini Travel Drive 64 GB USB 2.0 Flash ...   36.240         1   

      Discount    Profit  
2          0.0    6.8714  
5          0.0   14.1694  
6          0.0    1.9656  
7          0.2   90.7152  
8          0.2    5.7825  
...        ...       ...  
9976       0.2   31.1980  
9977       0.0    0.0000  
9978       0.2    4.6816  
9979       0.2  153.1152  
9986       0.0   15.2208  

[747 

### 👨‍💻 Task 8 - B

Pada DataFrame `data_superstore` tampilkan semua record data yang terjadi di 'Los Angeles' untuk kategori Technology

In [ ]:
# Write your code below here & exec!

# Assign conditions
conditions1 = (data_superstore['City'] == 'Los Angeles')
conditions2 = (data_superstore['Category'] == 'Technology')

# Filter
data_LA_tech = data_superstore[conditions1][conditions2]

# Tampilkeun
display(data_LA_tech)

<ipython-input-29-e7f1c8665658>:8: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  data_LA_tech = data_superstore[conditions1][conditions2]


Row ID        Order ID Order Date  Ship Date       Ship Mode  \
7          8  CA-2014-115812 2014-06-09 2014-06-14  Standard Class   
11        12  CA-2014-115812 2014-06-09 2014-06-14  Standard Class   
26        27  CA-2016-121755 2016-01-16 2016-01-20    Second Class   
62        63  CA-2015-135545 2015-11-24 2015-11-30  Standard Class   
90        91  CA-2016-109806 2016-09-17 2016-09-22  Standard Class   
...      ...             ...        ...        ...             ...   
9884    9885  CA-2014-112291 2014-04-03 2014-04-08  Standard Class   
9885    9886  CA-2014-112291 2014-04-03 2014-04-08  Standard Class   
9973    9974  US-2016-103674 2016-12-06 2016-12-10  Standard Class   
9976    9977  US-2016-103674 2016-12-06 2016-12-10  Standard Class   
9986    9987  CA-2016-125794 2016-09-29 2016-10-03  Standard Class   

     Customer ID    Customer Name      Segment        Country         City  \
7       BH-11710  Brosina Hoffman     Consumer  United States  Los Angeles   
11      BH-11710  Brosina Hoffman     Consumer  United States  Los Angeles   
26      EH-13945    Eric Hoffmann     Consumer  United States  Los Angeles   
62      KM-16720     Kunst Miller     Consumer  United States  Los Angeles   
90      JS-15685         Jim Sink    Corporate  United States  Los Angeles   
...          ...              ...          ...            ...          ...   
9884    KE-16420  Katrina Edelman    Corporate  United States  Los Angeles   
9885    KE-16420  Katrina Edelman    Corporate  United States  Los Angeles   
9973    AP-10720       Anne Pryor  Home Office  United States  Los Angeles   
9976    AP-10720       Anne Pryor  Home Office  United States  Los Angeles   
9986    ML-17410     Maris LaWare     Consumer  United States  Los Angeles   

      ... Postal Code  Region       Product ID    Category Sub-Category  \
7     ...       90032    West  TEC-PH-10002275  Technology       Phones   
11    ...       90032    West  TEC-PH-10002033  Technology       Phones   
26    ...       90049    West  TEC-AC-10003027  Technology  Accessories   
62    ...       90004    West  TEC-AC-10004633  Technology  Accessories   
90    ...       90036    West  TEC-PH-10004093  Technology       Phones   
...   ...         ...     ...              ...         ...          ...   
9884  ...       90008    West  TEC-AC-10001284  Technology  Accessories   
9885  ...       90008    West  TEC-AC-10000736  Technology  Accessories   
9973  ...       90032    West  TEC-PH-10004080  Technology       Phones   
9976  ...       90032    West  TEC-PH-10002496  Technology       Phones   
9986  ...       90008    West  TEC-AC-10003399  Technology  Accessories   

                                           Product Name    Sales  Quantity  \
7                        Mitel 5320 IP Phone VoIP phone  907.152         6   
11        Konftel 250 Conference phone - Charcoal black  911.424         4   
26     Imation 8GB Mini TravelDrive USB 2.0 Flash Drive   90.570         3   
62    Verbatim 25 GB 6x Blu-ray Single Layer Recorda...   13.980         2   
90                                   Panasonic Kx-TS550   73.584         2   
...                                                 ...      ...       ...   
9884  Enermax Briskie RF Wireless Keyboard and Mouse...   62.310         3   
9885                     Logitech G600 MMO Gaming Mouse  159.980         2   
9973                           Avaya 5410 Digital phone  271.960         5   
9976                                       Cisco SPA301  249.584         2   
9986  Memorex Mini Travel Drive 64 GB USB 2.0 Flash ...   36.240         1   

      Discount   Profit  
7          0.2  90.7152  
11         0.2  68.3568  
26         0.0  11.7741  
62         0.0   6.1512  
90         0.2   8.2782  
...        ...      ...  
9884       0.0  22.4316  
9885       0.0  57.5928  
9973       0.2  27.1960  
9976       0.2  31.1980  
9986       0.0  15.2208  

[150 rows x 21 columns]

In [ ]:
# Write your code below here & exec!

# Assign conditions
conditions1 = (data_superstore['City'] == 'Los Angeles')
conditions2 = (data_superstore['Category'] == 'Technology')

# Filter
data_LA_tech = data_superstore.loc[conditions1&conditions2]

# Tampilkeun
display(data_LA_tech)

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
7          8  CA-2014-115812 2014-06-09 2014-06-14  Standard Class   
11        12  CA-2014-115812 2014-06-09 2014-06-14  Standard Class   
26        27  CA-2016-121755 2016-01-16 2016-01-20    Second Class   
62        63  CA-2015-135545 2015-11-24 2015-11-30  Standard Class   
90        91  CA-2016-109806 2016-09-17 2016-09-22  Standard Class   
...      ...             ...        ...        ...             ...   
9884    9885  CA-2014-112291 2014-04-03 2014-04-08  Standard Class   
9885    9886  CA-2014-112291 2014-04-03 2014-04-08  Standard Class   
9973    9974  US-2016-103674 2016-12-06 2016-12-10  Standard Class   
9976    9977  US-2016-103674 2016-12-06 2016-12-10  Standard Class   
9986    9987  CA-2016-125794 2016-09-29 2016-10-03  Standard Class   

     Customer ID    Customer Name      Segment        Country         City  \
7       BH-11710  Brosina Hoffman     Consumer  United States  Los Angeles   
11      BH-11710  Brosina Hoffman     Consumer  United States  Los Angeles   
26      EH-13945    Eric Hoffmann     Consumer  United States  Los Angeles   
62      KM-16720     Kunst Miller     Consumer  United States  Los Angeles   
90      JS-15685         Jim Sink    Corporate  United States  Los Angeles   
...          ...              ...          ...            ...          ...   
9884    KE-16420  Katrina Edelman    Corporate  United States  Los Angeles   
9885    KE-16420  Katrina Edelman    Corporate  United States  Los Angeles   
9973    AP-10720       Anne Pryor  Home Office  United States  Los Angeles   
9976    AP-10720       Anne Pryor  Home Office  United States  Los Angeles   
9986    ML-17410     Maris LaWare     Consumer  United States  Los Angeles   

      ... Postal Code  Region       Product ID    Category Sub-Category  \
7     ...       90032    West  TEC-PH-10002275  Technology       Phones   
11    ...       90032    West  TEC-PH-10002033  Technology       Phones   
26    ...       90049    West  TEC-AC-10003027  Technology  Accessories   
62    ...       90004    West  TEC-AC-10004633  Technology  Accessories   
90    ...       90036    West  TEC-PH-10004093  Technology       Phones   
...   ...         ...     ...              ...         ...          ...   
9884  ...       90008    West  TEC-AC-10001284  Technology  Accessories   
9885  ...       90008    West  TEC-AC-10000736  Technology  Accessories   
9973  ...       90032    West  TEC-PH-10004080  Technology       Phones   
9976  ...       90032    West  TEC-PH-10002496  Technology       Phones   
9986  ...       90008    West  TEC-AC-10003399  Technology  Accessories   

                                           Product Name    Sales  Quantity  \
7                        Mitel 5320 IP Phone VoIP phone  907.152         6   
11        Konftel 250 Conference phone - Charcoal black  911.424         4   
26     Imation 8GB Mini TravelDrive USB 2.0 Flash Drive   90.570         3   
62    Verbatim 25 GB 6x Blu-ray Single Layer Recorda...   13.980         2   
90                                   Panasonic Kx-TS550   73.584         2   
...                                                 ...      ...       ...   
9884  Enermax Briskie RF Wireless Keyboard and Mouse...   62.310         3   
9885                     Logitech G600 MMO Gaming Mouse  159.980         2   
9973                           Avaya 5410 Digital phone  271.960         5   
9976                                       Cisco SPA301  249.584         2   
9986  Memorex Mini Travel Drive 64 GB USB 2.0 Flash ...   36.240         1   

      Discount   Profit  
7          0.2  90.7152  
11         0.2  68.3568  
26         0.0  11.7741  
62         0.0   6.1512  
90         0.2   8.2782  
...        ...      ...  
9884       0.0  22.4316  
9885       0.0  57.5928  
9973       0.2  27.1960  
9976       0.2  31.1980  
9986       0.0  15.2208  

[150 rows x 21 columns]

In [ ]:
# Write your code below here & exec!

# Assign conditions
conditions1 = (data_superstore['City'] == 'Los Angeles')
conditions2 = (data_superstore['Category'] == 'Technology')

# Filter
data_LA = data_superstore.loc[conditions1]

# Filter
data_LA_tech = data_LA.loc[conditions2]

# Tampilkeun
display(data_LA_tech)

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
7          8  CA-2014-115812 2014-06-09 2014-06-14  Standard Class   
11        12  CA-2014-115812 2014-06-09 2014-06-14  Standard Class   
26        27  CA-2016-121755 2016-01-16 2016-01-20    Second Class   
62        63  CA-2015-135545 2015-11-24 2015-11-30  Standard Class   
90        91  CA-2016-109806 2016-09-17 2016-09-22  Standard Class   
...      ...             ...        ...        ...             ...   
9884    9885  CA-2014-112291 2014-04-03 2014-04-08  Standard Class   
9885    9886  CA-2014-112291 2014-04-03 2014-04-08  Standard Class   
9973    9974  US-2016-103674 2016-12-06 2016-12-10  Standard Class   
9976    9977  US-2016-103674 2016-12-06 2016-12-10  Standard Class   
9986    9987  CA-2016-125794 2016-09-29 2016-10-03  Standard Class   

     Customer ID    Customer Name      Segment        Country         City  \
7       BH-11710  Brosina Hoffman     Consumer  United States  Los Angeles   
11      BH-11710  Brosina Hoffman     Consumer  United States  Los Angeles   
26      EH-13945    Eric Hoffmann     Consumer  United States  Los Angeles   
62      KM-16720     Kunst Miller     Consumer  United States  Los Angeles   
90      JS-15685         Jim Sink    Corporate  United States  Los Angeles   
...          ...              ...          ...            ...          ...   
9884    KE-16420  Katrina Edelman    Corporate  United States  Los Angeles   
9885    KE-16420  Katrina Edelman    Corporate  United States  Los Angeles   
9973    AP-10720       Anne Pryor  Home Office  United States  Los Angeles   
9976    AP-10720       Anne Pryor  Home Office  United States  Los Angeles   
9986    ML-17410     Maris LaWare     Consumer  United States  Los Angeles   

      ... Postal Code  Region       Product ID    Category Sub-Category  \
7     ...       90032    West  TEC-PH-10002275  Technology       Phones   
11    ...       90032    West  TEC-PH-10002033  Technology       Phones   
26    ...       90049    West  TEC-AC-10003027  Technology  Accessories   
62    ...       90004    West  TEC-AC-10004633  Technology  Accessories   
90    ...       90036    West  TEC-PH-10004093  Technology       Phones   
...   ...         ...     ...              ...         ...          ...   
9884  ...       90008    West  TEC-AC-10001284  Technology  Accessories   
9885  ...       90008    West  TEC-AC-10000736  Technology  Accessories   
9973  ...       90032    West  TEC-PH-10004080  Technology       Phones   
9976  ...       90032    West  TEC-PH-10002496  Technology       Phones   
9986  ...       90008    West  TEC-AC-10003399  Technology  Accessories   

                                           Product Name    Sales  Quantity  \
7                        Mitel 5320 IP Phone VoIP phone  907.152         6   
11        Konftel 250 Conference phone - Charcoal black  911.424         4   
26     Imation 8GB Mini TravelDrive USB 2.0 Flash Drive   90.570         3   
62    Verbatim 25 GB 6x Blu-ray Single Layer Recorda...   13.980         2   
90                                   Panasonic Kx-TS550   73.584         2   
...                                                 ...      ...       ...   
9884  Enermax Briskie RF Wireless Keyboard and Mouse...   62.310         3   
9885                     Logitech G600 MMO Gaming Mouse  159.980         2   
9973                           Avaya 5410 Digital phone  271.960         5   
9976                                       Cisco SPA301  249.584         2   
9986  Memorex Mini Travel Drive 64 GB USB 2.0 Flash ...   36.240         1   

      Discount   Profit  
7          0.2  90.7152  
11         0.2  68.3568  
26         0.0  11.7741  
62         0.0   6.1512  
90         0.2   8.2782  
...        ...      ...  
9884       0.0  22.4316  
9885       0.0  57.5928  
9973       0.2  27.1960  
9976       0.2  31.1980  
9986       0.0  15.2208  

[150 rows x 21 columns]

### 👨‍💻 Task 8 - C

Pada DataFrame `data_superstore` tampilkan record data yang terjadi di 'Los Angeles' untuk kategori Technology. Namun hanya tampilkan kolom 'Order_ID', 'City', 'Category' dan 'Profit'

In [ ]:
# Write your code below here & exec!

# Assign conditions
conditions1 = (data_superstore['City'] == 'Los Angeles')
conditions2 = (data_superstore['Category'] == 'Technology')
list_shown = ['Order ID', 'City', 'Category','Profit']

# Filter
data_LA_tech = data_superstore.loc[conditions1&conditions2, list_shown]

# Tampilkeun
display(data_LA_tech)

Order ID         City    Category   Profit
7     CA-2014-115812  Los Angeles  Technology  90.7152
11    CA-2014-115812  Los Angeles  Technology  68.3568
26    CA-2016-121755  Los Angeles  Technology  11.7741
62    CA-2015-135545  Los Angeles  Technology   6.1512
90    CA-2016-109806  Los Angeles  Technology   8.2782
...              ...          ...         ...      ...
9884  CA-2014-112291  Los Angeles  Technology  22.4316
9885  CA-2014-112291  Los Angeles  Technology  57.5928
9973  US-2016-103674  Los Angeles  Technology  27.1960
9976  US-2016-103674  Los Angeles  Technology  31.1980
9986  CA-2016-125794  Los Angeles  Technology  15.2208

[150 rows x 4 columns]

In [ ]:
# Write your code below here & exec!

# Assign conditions
conditions1 = (data_superstore['City'] == 'Los Angeles')
conditions2 = (data_superstore['Category'] == 'Technology')
list_shown = ['Order ID', 'City', 'Category','Profit']

# Filter
data_LA = data_superstore.loc[conditions1]

# Filter
data_LA_tech = data_LA.loc[conditions2][list_shown]

# Tampilkeun
display(data_LA_tech)

Order ID         City    Category   Profit
7     CA-2014-115812  Los Angeles  Technology  90.7152
11    CA-2014-115812  Los Angeles  Technology  68.3568
26    CA-2016-121755  Los Angeles  Technology  11.7741
62    CA-2015-135545  Los Angeles  Technology   6.1512
90    CA-2016-109806  Los Angeles  Technology   8.2782
...              ...          ...         ...      ...
9884  CA-2014-112291  Los Angeles  Technology  22.4316
9885  CA-2014-112291  Los Angeles  Technology  57.5928
9973  US-2016-103674  Los Angeles  Technology  27.1960
9976  US-2016-103674  Los Angeles  Technology  31.1980
9986  CA-2016-125794  Los Angeles  Technology  15.2208

[150 rows x 4 columns]

In [ ]:
# Write your code below here & exec!

# Assign conditions
conditions1 = (data_superstore['City'] == 'Los Angeles')
conditions2 = (data_superstore['Category'] == 'Technology')
list_shown = ['Order ID', 'City', 'Category','Profit']

# Filter
data_LA_tech = data_superstore[conditions1][conditions2][list_shown]

# Tampilkeun
display(data_LA_tech)

<ipython-input-51-8d8298f18a91>:9: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  data_LA_tech = data_superstore[conditions1][conditions2][list_shown]


Order ID         City    Category   Profit
7     CA-2014-115812  Los Angeles  Technology  90.7152
11    CA-2014-115812  Los Angeles  Technology  68.3568
26    CA-2016-121755  Los Angeles  Technology  11.7741
62    CA-2015-135545  Los Angeles  Technology   6.1512
90    CA-2016-109806  Los Angeles  Technology   8.2782
...              ...          ...         ...      ...
9884  CA-2014-112291  Los Angeles  Technology  22.4316
9885  CA-2014-112291  Los Angeles  Technology  57.5928
9973  US-2016-103674  Los Angeles  Technology  27.1960
9976  US-2016-103674  Los Angeles  Technology  31.1980
9986  CA-2016-125794  Los Angeles  Technology  15.2208

[150 rows x 4 columns]

## 📎 **9. Sorting Row**

Menemukan data yang menarik dalam sebuah DataFrame seringkali lebih mudah jika dilakukan pengurutan baris datanya. Pengurutan dibagi menjadi dua yakni pengurutan secara `Ascending` (pengurutan naik A → Z) dan juga pengurutan secara `Descending` (pengurutan turun Z → A). Eksekusi sintaks berikut :

<p align="center"><code><b>DataFrame.sort_values(by, ascending=True, ignore_index=False)</b></code></p>

Jika ingin mengurutkan DataFrame pandas dengan memperhatikan argumen

<ul>
<li><b><i>by</i></b> sebagai satu atau lebih kolom yang ingin dijadikan patokan pengurutan</li>

<li><b><i>ascending</i></b> dapat diisikan True jika ingin terurut naik atau False jika ingin terurut turun (dafault : True)</li>

<li><b><i>ignore_index</i></b> dapat diisikan True jika ingin index masih terurut atau False jika indexnya masih dipertahankan sesuai data awal (dafault : False)</li>

</ul><br>

docs : <i>https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html</i>


### 👨‍💻 Task 9 - A

Pada DataFrame `data_superstore`, tampilkan semua record data namun urutkan secara ascending pada kolom Order_ID

In [ ]:
# Write your code below here & exec!
data_superstore.sort_values(by = ['Order ID'], ascending=True, ignore_index=True)

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
0       2718  CA-2014-100006 2014-09-07 2014-09-13  Standard Class   
1       6289  CA-2014-100090 2014-07-08 2014-07-12  Standard Class   
2       6288  CA-2014-100090 2014-07-08 2014-07-12  Standard Class   
3       9515  CA-2014-100293 2014-03-14 2014-03-18  Standard Class   
4       3084  CA-2014-100328 2014-01-28 2014-02-03  Standard Class   
...      ...             ...        ...        ...             ...   
9989    5933  US-2017-169551 2017-07-07 2017-07-09     First Class   
9990    5934  US-2017-169551 2017-07-07 2017-07-09     First Class   
9991    5935  US-2017-169551 2017-07-07 2017-07-09     First Class   
9992    5930  US-2017-169551 2017-07-07 2017-07-09     First Class   
9993    5932  US-2017-169551 2017-07-07 2017-07-09     First Class   

     Customer ID     Customer Name      Segment        Country           City  \
0       DK-13375       Dennis Kane     Consumer  United States  New York City   
1       EB-13705        Ed Braxton    Corporate  United States  San Francisco   
2       EB-13705        Ed Braxton    Corporate  United States  San Francisco   
3       NF-18475  Neil Französisch  Home Office  United States   Jacksonville   
4       JC-15340   Jasper Cacioppo     Consumer  United States  New York City   
...          ...               ...          ...            ...            ...   
9989    RL-19615         Rob Lucas     Consumer  United States   Philadelphia   
9990    RL-19615         Rob Lucas     Consumer  United States   Philadelphia   
9991    RL-19615         Rob Lucas     Consumer  United States   Philadelphia   
9992    RL-19615         Rob Lucas     Consumer  United States   Philadelphia   
9993    RL-19615         Rob Lucas     Consumer  United States   Philadelphia   

      ... Postal Code  Region       Product ID         Category Sub-Category  \
0     ...       10024    East  TEC-PH-10002075       Technology       Phones   
1     ...       94122    West  OFF-BI-10001597  Office Supplies      Binders   
2     ...       94122    West  FUR-TA-10003715        Furniture       Tables   
3     ...       32216   South  OFF-PA-10000176  Office Supplies        Paper   
4     ...       10024    East  OFF-BI-10000343  Office Supplies      Binders   
...   ...         ...     ...              ...              ...          ...   
9989  ...       19120    East  OFF-ST-10004835  Office Supplies      Storage   
9990  ...       19120    East  TEC-AC-10002018       Technology  Accessories   
9991  ...       19120    East  TEC-AC-10003033       Technology  Accessories   
9992  ...       19120    East  FUR-BO-10001519        Furniture    Bookcases   
9993  ...       19120    East  TEC-PH-10001363       Technology       Phones   

                                           Product Name    Sales  Quantity  \
0                                     AT&T EL51110 DECT  377.970         3   
1     Wilson Jones Ledger-Size, Piano-Hinge Binder, ...  196.704         6   
2               Hon 2111 Invitation Series Corner Table  502.488         3   
3                                            Xerox 1887   91.056         6   
4     Pressboard Covers with Storage Hooks, 9 1/2" x...    3.928         1   
...                                                 ...      ...       ...   
9989                  Plastic Stacking Crates & Casters   13.392         3   
9990              AmazonBasics 3-Button USB Wired Mouse   16.776         3   
9991  Plantronics CS510 - Over-the-Head monaural Wir...  527.920         2   
9992            O'Sullivan 3-Shelf Heavy-Duty Bookcases   87.210         3   
9993                                    Apple iPhone 5S  683.988         2   

      Discount    Profit  
0          0.0  109.6113  
1          0.2   68.8464  
2          0.2  -87.9354  
3          0.2   31.8696  
4          0.2    1.3257  
...        ...       ...  
9989       0.2    1.0044  
9990       0.2    4.8231  
9991       0.2   85.7870  
9992       0.5  -45.3492

### 👨‍💻 Task 9 - B

Pada DataFrame `data_superstore`, tampilkan semua record data namun urutkan secara descending pada kolom Profit dan urutkan secara ascending pada kolom Quantity

In [ ]:
# Write your code below here & exec!
sorted_data = data_superstore.sort_values(by = ['Profit','Quantity'],ascending=[False, True])

display(sorted_data)

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
6826    6827  CA-2016-118689 2016-10-02 2016-10-09  Standard Class   
8153    8154  CA-2017-140151 2017-03-23 2017-03-25     First Class   
4190    4191  CA-2017-166709 2017-11-17 2017-11-22  Standard Class   
9039    9040  CA-2016-117121 2016-12-17 2016-12-21  Standard Class   
4098    4099  CA-2014-116904 2014-09-23 2014-09-28  Standard Class   
...      ...             ...        ...        ...             ...   
4991    4992  US-2017-122714 2017-12-07 2017-12-13  Standard Class   
3011    3012  CA-2017-134845 2017-04-17 2017-04-23  Standard Class   
9774    9775  CA-2014-169019 2014-07-26 2014-07-30  Standard Class   
683      684  US-2017-168116 2017-11-04 2017-11-04        Same Day   
7772    7773  CA-2016-108196 2016-11-25 2016-12-02  Standard Class   

     Customer ID   Customer Name      Segment        Country         City  \
6826    TC-20980    Tamara Chand    Corporate  United States    Lafayette   
8153    RB-19360    Raymond Buch     Consumer  United States      Seattle   
4190    HL-15040    Hunter Lopez     Consumer  United States       Newark   
9039    AB-10105   Adrian Barton     Consumer  United States      Detroit   
4098    SC-20095    Sanjit Chand     Consumer  United States  Minneapolis   
...          ...             ...          ...            ...          ...   
4991    HG-14965   Henry Goldwyn    Corporate  United States      Chicago   
3011    SR-20425  Sharelle Roach  Home Office  United States   Louisville   
9774    LF-17185     Luke Foster     Consumer  United States  San Antonio   
683     GT-14635  Grant Thornton    Corporate  United States   Burlington   
7772    CS-12505   Cindy Stewart     Consumer  United States    Lancaster   

      ... Postal Code   Region       Product ID         Category Sub-Category  \
6826  ...       47905  Central  TEC-CO-10004722       Technology      Copiers   
8153  ...       98115     West  TEC-CO-10004722       Technology      Copiers   
4190  ...       19711     East  TEC-CO-10004722       Technology      Copiers   
9039  ...       48205  Central  OFF-BI-10000545  Office Supplies      Binders   
4098  ...       55407  Central  OFF-BI-10001120  Office Supplies      Binders   
...   ...         ...      ...              ...              ...          ...   
4991  ...       60653  Central  OFF-BI-10001120  Office Supplies      Binders   
3011  ...       80027     West  TEC-MA-10000822       Technology     Machines   
9774  ...       78207  Central  OFF-BI-10004995  Office Supplies      Binders   
683   ...       27217    South  TEC-MA-10004125       Technology     Machines   
7772  ...       43130     East  TEC-MA-10000418       Technology     Machines   

                                          Product Name      Sales  Quantity  \
6826             Canon imageCLASS 2200 Advanced Copier  17499.950         5   
8153             Canon imageCLASS 2200 Advanced Copier  13999.960         4   
4190             Canon imageCLASS 2200 Advanced Copier  10499.970         3   
9039  GBC Ibimaster 500 Manual ProClick Binding System   9892.740        13   
4098              Ibico EPK-21 Electric Binding System   9449.950         5   
...                                                ...        ...       ...   
4991              Ibico EPK-21 Electric Binding System   1889.990         5   
3011         Lexmark MX611dhe Monochrome Laser Printer   2549.985         5   
9774         GBC DocuBind P400 Electric Binding System   2177.584         8   
683          Cubify CubeX 3D Printer Triple Head Print   7999.980         4   
7772         Cubify CubeX 3D Printer Double Head Print   4499.985         5   

      Discount     Profit  
6826       0.0  8399.9760  
8153       0.0  6719.9808  
4190       0.0  5039.9856  
9039       0.0  4946.3700  
4098       0.0  4630.4755  
...        ...        ...  
4991       0.8 -2929.4845  
3011       0.7 -3399.9800  
9774       0.8 -3701.8928  
683        0.5 -3839.9904  
7772      

## 📊 **10. Summarize Data**

Di pandas, method `DataFrame.describe()` digunakan untuk menghasilkan statistik deskriptif dari DataFrame. Metode ini sangat berguna untuk dengan cepat memahami statistik dasar dari data. Berikut adalah statistik utama yang disediakan :

* `Jumlah` : Jumlah nilai non-null dalam setiap kolom.
* `Rata-rata` : Rata-rata aritmetika dari setiap kolom.
* `Standar Deviasi` : Ukuran dispersi nilai-nilai sekitar rata-rata.
* `Minimum` : Nilai minimum dalam setiap kolom.
* `Kuartil 25 (Q1)` : Nilai di bawahnya 25% data berada.
* `Kuartil 50 (Median atau Q2)` : Nilai di bawahnya 50% data berada (yaitu nilai tengah).
* `Kuartil 75 (Q3)` : Nilai di bawahnya 75% data berada.
* `Maksimum` : Nilai maksimum dalam setiap kolom.

### 👨‍💻 Task 10

Pada DataFrame `data_superstore`, lakukan perhitungan statistik deskriptif pada kolom - kolom yang memiliki tipe data numerik

In [ ]:
# Write your code below here & exec!
data_superstore.describe()

Row ID                     Order Date  \
count  9994.000000                           9994   
mean   4997.500000  2016-04-30 00:07:12.259355648   
min       1.000000            2014-01-03 00:00:00   
25%    2499.250000            2015-05-23 00:00:00   
50%    4997.500000            2016-06-26 00:00:00   
75%    7495.750000            2017-05-14 00:00:00   
max    9994.000000            2017-12-30 00:00:00   
std    2885.163629                            NaN   

                           Ship Date   Postal Code         Sales     Quantity  \
count                           9994   9994.000000   9994.000000  9994.000000   
mean   2016-05-03 23:06:58.571142912  55190.379428    229.858001     3.789574   
min              2014-01-07 00:00:00   1040.000000      0.444000     1.000000   
25%              2015-05-27 00:00:00  23223.000000     17.280000     2.000000   
50%              2016-06-29 00:00:00  56430.500000     54.490000     3.000000   
75%              2017-05-18 00:00:00  90008.000000    209.940000     5.000000   
max              2018-01-05 00:00:00  99301.000000  22638.480000    14.000000   
std                              NaN  32063.693350    623.245101     2.225110   

          Discount       Profit  
count  9994.000000  9994.000000  
mean      0.156203    28.656896  
min       0.000000 -6599.978000  
25%       0.000000     1.728750  
50%       0.200000     8.666500  
75%       0.200000    29.364000  
max       0.800000  8399.976000  
std       0.206452   234.260108

In [ ]:
# Write your code below here & exec!
data_superstore.select_dtypes(include = 'number').describe().drop(columns = ['Row ID','Postal Code'])

Sales     Quantity     Discount       Profit
count   9994.000000  9994.000000  9994.000000  9994.000000
mean     229.858001     3.789574     0.156203    28.656896
std      623.245101     2.225110     0.206452   234.260108
min        0.444000     1.000000     0.000000 -6599.978000
25%       17.280000     2.000000     0.000000     1.728750
50%       54.490000     3.000000     0.200000     8.666500
75%      209.940000     5.000000     0.200000    29.364000
max    22638.480000    14.000000     0.800000  8399.976000

## 🎎 **11. Aggregation**

Agregasi data merupakan serangkaian proses yang dilakukan untuk melakukan pengumpulan data dan dinyatakan dalam bentuk ringkasan sebagai contoh menghitung total yang dikelompokkan berdasarkan kategori tertentu, dsb.

Untuk melakukan agregasi gunakan sintaks berikut

<p align="center"><code><b>DataFrame.group_by(by, as_index).agg(kolom_hasil_aggregasi = (kolom_dihitung, jenis_agregasi))</b></code></p>

Jenis agregasi :
* pd.Series.sum = Menghitung total data
* pd.Series.mean = Menghitung rataan data
* pd.Series.max = Menghitung nilai maksimum data
* pd.Series.min = Menghitung nilai minimum data
* pd.Series.nunique = Menghitung jumlah data unik data
* pd.Series.count = Menghitung banyak data

### 👨‍💻 Task 11 - A

Pada DataFrame `data_superstore`, Category produk apakah yang menghasilkan total profit paling tinggi? dan berapa total profit yang didapatkan?

In [ ]:
data_superstore.head(1)

Row ID        Order ID Order Date  Ship Date     Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   

  Customer Name   Segment        Country       City  ... Postal Code  Region  \
0   Claire Gute  Consumer  United States  Henderson  ...       42420   South   

        Product ID   Category Sub-Category                       Product Name  \
0  FUR-BO-10001798  Furniture    Bookcases  Bush Somerset Collection Bookcase   

    Sales  Quantity  Discount   Profit  
0  261.96         2       0.0  41.9136  

[1 rows x 21 columns]

In [ ]:
# Write your code below here & exec!
data_superstore.groupby(['Category']).agg(total_profit = ('Profit', 'sum'))

total_profit
Category                     
Furniture          18451.2728
Office Supplies   122490.8008
Technology        145454.9481

In [67]:
# NAMA KOLOM GABISA DICUSTOM

data_superstore.groupby(['Category'])['Profit'].sum()

Category
Furniture           18451.2728
Office Supplies    122490.8008
Technology         145454.9481
Name: Profit, dtype: float64

Technology menghasilkan total profit paling tinggi dengan total profit yang didapatkan 145454.9481

### 👨‍💻 Task 11 - B

Pada DataFrame `data_superstore`, lakukan perhitungan rataan Profit yang dikelompokkan berdasarkan Segment dan Category barang yang terjual

In [ ]:
data_superstore.head(1)

Row ID        Order ID Order Date  Ship Date     Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   

  Customer Name   Segment        Country       City  ... Postal Code  Region  \
0   Claire Gute  Consumer  United States  Henderson  ...       42420   South   

        Product ID   Category Sub-Category                       Product Name  \
0  FUR-BO-10001798  Furniture    Bookcases  Bush Somerset Collection Bookcase   

    Sales  Quantity  Discount   Profit  
0  261.96         2       0.0  41.9136  

[1 rows x 21 columns]

In [ ]:
# Write your code below here & exec!
avg_profit_per_category = data_superstore.groupby(['Segment','Category']).agg(rataan_profit = ('Profit', 'mean'))

display(avg_profit_per_category)

rataan_profit
Segment     Category                      
Consumer    Furniture             6.281293
            Office Supplies      18.014174
            Technology           74.445646
Corporate   Furniture            11.741201
            Office Supplies      22.102923
            Technology           79.723823
Home Office Furniture            10.705465
            Office Supplies      24.034439
            Technology           89.152458

In [68]:
# Write your code below here & exec!
avg_profit_per_category = data_superstore.groupby(['Segment','Category'], as_index = False).agg(rataan_profit = ('Profit', 'mean'))

display(avg_profit_per_category)

Segment         Category  rataan_profit
0     Consumer        Furniture       6.281293
1     Consumer  Office Supplies      18.014174
2     Consumer       Technology      74.445646
3    Corporate        Furniture      11.741201
4    Corporate  Office Supplies      22.102923
5    Corporate       Technology      79.723823
6  Home Office        Furniture      10.705465
7  Home Office  Office Supplies      24.034439
8  Home Office       Technology      89.152458

**Written By :** <br>
<a href="https://www.linkedin.com/in/azzah-m-165867214/"><img alt="Linked In Link" src="https://img.shields.io/badge/-Azzah%20M-0072b1?style=for-the-badge&logo=None&logoColor=white" align="left"/></a>

<br><br>**for :**

<a href="https://dqlab.id/"><img src="https://dqlab.id/files/dqlab/cache/87e30118ebba5ec7d96f6ea8c9dcc10b_x_118_X_55.png" align="left" /></a>